lib/src/extend/functions.dart

```diff
@@ -106,19 +104,63 @@ List<ComplexSelector>? unifyComplex(
 /// Returns a [CompoundSelector] that matches only elements that are matched by
 /// both [compound1] and [compound2].
 ///
-/// The [span] will be used for the new unified selector.
+/// The [compound1]'s `span` will be used for the new unified selector.
+///
+/// This function ensures that the relative order of pseudo-classes (`:`) and
+/// pseudo-elements (`::`) within each input selector is preserved in the
+/// resulting combined selector.
+///
+/// This function enforces a general rule that pseudo-classes (`:`) should come
+/// before pseudo-elements (`::`), but it won't change their order if they were
+/// originally interleaved within a single input selector. This prevents
+/// unintended changes to the selector's meaning. For example, unifying
+/// `::foo:bar` and `:baz` results in `:baz::foo:bar`. `:baz` is a pseudo-class,
+/// so it is moved before the pseudo-class `::foo`. Meanwhile, `:bar` is not
+/// moved before `::foo` because it appeared after `::foo` in the original
+/// selector.
 ///
 /// If no such selector can be produced, returns `null`.
 CompoundSelector? unifyCompound(
     CompoundSelector compound1, CompoundSelector compound2) {
-  var result = compound2.components;
-  for (var simple in compound1.components) {
-    var unified = simple.unify(result);
-    if (unified == null) return null;
-    result = unified;
+  var result = <SimpleSelector>[];
+  var pseudoResult = <SimpleSelector>[];
+  var pseudoElementFound = false;
+
+  // Compound selectors may not contain more than one pseudo-element selectors.
+  if (compound1.components
+          .followedBy(compound2.components)
+          .whereType<PseudoSelector>()
+          .where((pseudo) => pseudo.isElement)
+          .toSet()
+          .length >
+      1) {
+    return null;
+  }
+
+  for (var simple in [...compound1.components, null, ...compound2.components]) {
+    // The relative order of pseudo selectors is preserved only in the original
+    // selector. We use `null` as a signal to know when we transition from the
+    // first components into the second components.
+    if (simple == null) {
+      pseudoElementFound = false;
+      continue;
+    }
+
+    // All pseudo-classes are unified separately after a pseudo-element to
+    // preserve their relative order with the pseudo-element.
+    if (pseudoElementFound && simple is PseudoSelector) {
+      var unified = simple.unify(pseudoResult);
```

Comment:
I'm really nervous about the way this _re-runs_ unification for `compound2`, both from a performance and a correctness perspective. For example:

```scss
.foo.foo.bar {x: y}
.baz {@extend .bar}
```

now produces

```css
.foo.foo.bar, .foo.baz {
  x: y;
}
```

with the duplicated `.foo` removed. I think a better way to handle it would be to ensure that `simple.unify()` injects itself in the correct place relative to an existing pseudo-element.

Reply:
When I ran this before, I noticed that there were duplicate selectors, and kinda forgot about specificity! I'll change this back :)